### DISCUSSION TITLE
Ruby/Furigana Support

### DISCUSSION THREAD

**Discussion by MichaelCharles** [marp-team/marp#416] [Ideas] [answered]
I think it would be great if we had support for ruby, or as it is called in Japanese furigana.

Furigana are basically phonetic guide words that appear above written text, such as in the image below.

<img src="https://res.cloudinary.com/mca62511/image/upload/v1675672889/furigana_oyueke.png" width="200px" />

On the r/learnjapanese community on Reddit, for a long time they had a syntax for furigana in Markdown that made a lot of sense to me. It looked like this,

```
このように[振り仮名](#fg "ふりがな")を[使](#fg "つか")えます！
```

I'm currently writing a presentation in Japanese, and it would be really helpful it furigana were supported.

I know other languages use ruby/phonetic guides as well, so it would be potentially helpful even to those who aren't using Japanese. 

Though I guess I can just do it in inline HTML?

```
<p>このように<ruby>振り仮名<rp>(</rp><rt>ふりがな</rt><rp>)</rp></ruby>を<ruby>使<rp>(</rp><rt>つか</rt><rp>)</rp></ruby>えます</p>
```

But that's kind of ugly and difficult to parse when editing.

**Comment by yhatt** [comment, marked as answer]
Marp's [plugin architecture](https://marpit.marp.app/usage?id=extend-marpit-by-plugins) is just for that. You can add a custom furigana syntax to Marp by using markdown-it plugins: https://www.npmjs.com/search?q=markdown-it%20ruby

You must use [Marp CLI](https://github.com/marp-team/marp-cli#functional-engine) whenever using any plugins.

### Example: [`markdown-it-ruby`](https://github.com/lostandfound/markdown-it-ruby)

```markdown
`markdown-it-ruby` を{使|つか}うと、このような{構文|こうぶん}を{使用|しよう}して、{漢字|かんじ}にルビを{振|ふ}ることができます。

Marp は{汎用性|はんようせい}の{高|たか}いエコシステムとして{設計|せっけい}されているため、CommonMark {外|がい}の Markdown {構文|こうぶん}に{関|かん}して、{独自|どくじ}の{意見|いけん}を{持|も}つことを{極力|きょくりょく}{避|さ}けています。{必要|ひつよう}に{応|おう}じて、ユーザーがプラグインを{使|つか}うことで Markdown {構文|こうぶん}が{拡張|かくちょう}されるべきです。
```

```javascript
// engine.js
module.exports = ({ marp }) => marp.use(require('markdown-it-ruby'));
```

```bash
npm i --save markdown-it-ruby
marp --engine ./engine.js japanese.md
```

![スクリーンショット 2023-02-06 22 19 58](https://user-images.githubusercontent.com/3993388/216981848-05681057-9e0e-4fce-88f6-92ac01267e83.png)

[There are various suggestions about the furigana syntax in Markdown](https://talk.commonmark.org/t/proper-ruby-text-rb-syntax-support-in-markdown), and `markdown-it-ruby` is using the [DenDenMarkdown](https://conv.denshochan.com/markdown#ruby) style syntax. If you want the furigana support with the syntax you've described, you may have to write another markdown-it plugin: https://github.com/markdown-it/markdown-it/blob/master/docs/architecture.md
